### PR TITLE
Create Update Profile Info/Set Contacts form

### LINKED ISSUE #70: Edit User and Contacts information
[Project/Issue Tracker](https://github.com/Code-4-Community/lucys-love-bus-frontend/projects/1)

### Objective 🦋 

As a user I want to go to a page that contains a form autofilled with my current user, contacts, and children information. I want to be able to modify all fields (besides password and email) and send a API request to update all my contact information.
 
### Tech Spec

- [ ] Create a single page form for filling out all contact information plus relevant user information
  - [ ] Autofills with current data, aka pull the current data and auto populate the form fields with that information
  - [ ] Includes the photo release field from the user table
  - [ ] Includes all contact info
    - [ ] Main contact (the account owner)
    - [ ] List of additional contacts
    - [ ] List of children 

### Notes

The child and additional contact forms are already completed as components from the sign up flow, so we can simply reuse those.

### PR DESCRIPTION
**Pull request by rymaju**
## Issue

Closes #70 
Linked backend PR: https://github.com/Code-4-Community/lucys-love-bus-backend/pull/125

## Changes

- Created Update Profile Info page/Set Contacts PF signup flow page (same thing)
- Made various form fragment components
- Created redux files for contacts
- Linked signup PF form to set contacts page
- Refactored code into signup confirmation page which is now its own container, signup routes there if GM, otherwise set contacts when in the PF signup flow.

## Screenshots

image

## Verification Steps

- Looked in pgadmin to see if children and contacts were properly created.
- Children and contacts were created with correct S3 image links.
- Was able to update contacts multiple times, including photos which updated properly.

## Notes

Backend PR https://github.com/Code-4-Community/lucys-love-bus-backend/pull/125 is required for this ticket to work in production

### CODE REVIEW

**Comment by rymaju** [comment]
Note to reviewers: I got burnt out approximately 700 lines into this PR so I guarantee there is a lot of suspicious code that can be improved here. I went ahead and commented in places I think need another pair of eyes 👁️ 👃 👁️

File: src/api/protectedApiClient.ts [hunk 1]
```diff
@@ -13,13 +14,16 @@ export interface ProtectedApiClient {
   readonly getRequestStatuses: () => Promise<PersonalRequest[]>;
   readonly makePFRequest: () => Promise<void>;
   readonly deactivateAccount: () => Promise<void>;
+  readonly getContactInfo: () => Promise<ContactInfo>;
+  readonly setContactInfo: (request: ContactInfo) => Promise<void>;
```
Comment (rymaju) [inline src/api/protectedApiClient.ts:18]: Why put this in redux?

File: src/api/protectedApiClient.ts [hunk 2]
```diff
@@ -40,6 +44,16 @@ const deactivateAccount = (): Promise<void> => {
     .catch((err) => err);
```
Comment (rymaju) [inline src/api/protectedApiClient.ts:44]: Not related to my PR but why are we catching errors here? It will be impossible to catch them in the thunk so we will get a "Completed" action with an error response

File: src/components/ChildFormFragment.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+const ChildFormFragment: React.FC<{ field: FormListFieldData }> = ({
+  field,
+}) => {
+  return (
+    <>
+      <Form.Item
+        name={[field.name, 'firstName']}
+        fieldKey={[field.fieldKey, 'firstName']}
+        label="First Name"
+        className="inline-block-half"
```
Comment (rymaju) [inline src/components/ChildFormFragment.tsx:16]: Why class names? These dont make sense

File: src/components/LocationFormFragment.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
+import { Form, Input } from 'antd';
+import React from 'react';
+
+// TODO: use this fragment in signup form, note that "zip" was renamed to "zipCode"
```
Comment (rymaju) [inline src/components/LocationFormFragment.tsx:4]: Leaving TODOs in code? I think in this case its ok because refactoring signup form should be its own PR

File: src/components/LocationFormFragment.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        <Input placeholder="State" />
+      </Form.Item>
+      <Form.Item
+        name="zipCode"
+        label="Zip Code"
+        className="block-third"
+        rules={[
+          {
+            required: true,
+            pattern: new RegExp(/^[^0-9]*(?:(\d)[^0-9]*){5}$/),
```
Comment (rymaju) [inline src/components/LocationFormFragment.tsx:45]: We need to test all our regex

File: src/containers/setContacts/ducks/thunks.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    dispatch(setContacts.loading());
+    return protectedApiClient
+      .setContactInfo(request)
+      .then(() => {
+        dispatch(contacts.loaded(request));
+        dispatch(setContacts.loaded());
+      })
+      .catch((error) => {
+        dispatch(contacts.failed(error));
+        dispatch(setContacts.failed(error));
```
Comment (rymaju) [inline src/containers/setContacts/ducks/thunks.ts:32]: Dispatch to contacts too? is this redundant given that I load contacts with a useEffect on the contacts form page?

File: src/containers/setContacts/index.tsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+const SetContacts: React.FC<SetContactsProps> = ({ contacts, setContacts }) => {
+  const dispatch = useDispatch();
+  const history = useHistory();
+  const location = useLocation();
+  useEffect(() => {
+    dispatch(getContactInfo());
+  }, [dispatch]);
+  const registeringAsParticipatingFamily =
```
Comment (rymaju) [inline src/containers/setContacts/index.tsx:44]: Abstract to utils?

File: src/containers/setContacts/index.tsx [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+  if (asyncRequestIsComplete(setContacts) && registeringAsParticipatingFamily) {
+    protectedApiClient.makePFRequest();
+    history.push({
+      pathname: Routes.SIGNUP_CONFIRMATION,
+      search: participatingFamilySearchQuery,
+    });
+  }
+
+  function selectProfilePictureOrDefaultToUndefined(
```
Comment (rymaju) [inline src/containers/setContacts/index.tsx:57]: Does this need to be moved to utils? It would only ever be used right here. Needs jsdoc to explain what it does? Better naming?

File: src/containers/setContacts/index.tsx [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+      return undefined;
+    }
+  }
+
+  const handleCloseSuccessAlert = () => {
+    dispatch(resetSetContactState());
+  };
+
+  const onFinish = async (values: ContactFormFields) => {
+    if (!asyncRequestIsComplete(contacts)) {
```
Comment (rymaju) [inline src/containers/setContacts/index.tsx:73]: What should be done here? Throwing an error will do nothing I think. Currently error is linked to the contacts redux, maybe we can use useState and have the error display if a custom error or the redux error appears?

File: src/containers/setContacts/index.tsx [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+          address: values.address,
+          city: values.city,
+          state: values.state,
+          zipCode: values.zipCode,
+        },
+      }),
+    );
+  };
+
+  const contactsToFormFields = (info: ContactInfo): ContactFormFields => {
```
Comment (rymaju) [inline src/containers/setContacts/index.tsx:137]: Does this require documentation or more context?

File: src/containers/setContacts/index.tsx [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+    return {
+      ...info.location,
+      ...info.mainContact,
+      dateOfBirth: moment(new Date()),
+      profilePicture: undefined,
+      additionalContacts:
+        info.additionalContacts &&
+        info.additionalContacts.map((c) => ({
+          ...c,
+          dateOfBirth: moment(new Date()),
```
Comment (rymaju) [inline src/containers/setContacts/index.tsx:147]: imagine importing 290k of code for date utility functions 😔

File: src/utils/fileEncoding.ts [hunk 12]
```diff
@@ -19,14 +19,14 @@ export function convertToBase64String(file: File): Promise<string> {
   });
 }
 
-interface FileField {
+export interface FileField {
   file: File;
 }
 
 export const encodeProfileFieldFile = async (
```
Comment (rymaju) [inline src/utils/fileEncoding.ts:26]: Does this need jsdoc?

Dont know how to test these functions that require file objects.

File: src/components/ChildFormFragment.tsx (hunk 3 above)
Comment (anirudh723) [reply src/components/ChildFormFragment.tsx:16]: I agree, this does not make sense

File: src/components/LocationFormFragment.tsx (hunk 5 above)
Comment (anirudh723) [reply src/components/LocationFormFragment.tsx:45]: Facts ^

File: src/containers/setContacts/index.tsx [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+      throw new Error(
+        'Setting Form values requires knowing current contact information',
+      );
+    }
+
+    dispatch(
+      setContactInfo({
+        mainContact: {
+          email: contacts.result.mainContact.email,
+          firstName: values.firstName,
```
Comment (anirudh723) [inline src/containers/setContacts/index.tsx:83]: Maybe deconstruct `values` and pull out all the properties that you don't need, like:
`const {notNeeded1, notNeeded2, notNeeded3, ...requiredValues} = values;`

so then, you could do:
```
mainContact: {
   ...requiredValues,
   ...
}
```

I'm not sure how much this would help because for some of them properties of `values` you do conditionals, like for for pronouns, allergies, and diagnosis.

File: src/containers/setContacts/index.tsx (hunk 10 above)
Comment (anirudh723) [reply src/containers/setContacts/index.tsx:137]: maybe `mapContactToFormFields`

File: src/components/LocationFormFragment.tsx [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+            message: 'Please input a valid five-digit zip code',
```
Comment (anirudh723) [inline src/components/LocationFormFragment.tsx:46]: Are users only from the US? Can we have Canadian users for example, who have a slightly different zip code?

File: src/components/LocationFormFragment.tsx [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+        className="stacked-inputs"
+        rules={[{ required: true, message: 'Please input your address' }]}
+      >
+        <Input placeholder="Address" />
+      </Form.Item>
+      <Form.Item
+        name="city"
+        label="City"
+        className="inline-block-third"
+        rules={[{ required: true, message: 'Please input city' }]}
```
Comment (anirudh723) [inline src/components/LocationFormFragment.tsx:20]: This could be limited to strings only

File: src/components/RegistrationFormBody.tsx [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+import { FormListFieldData } from 'antd/lib/form/FormList';
+import React from 'react';
+const { Dragger } = Upload;
+const { TextArea } = Input;
+
+const RegistrationFormBody: React.FC<{
+  isMainContact?: boolean;
+  field?: FormListFieldData;
+}> = ({ isMainContact, field }) => {
+  function generateName(name: string) {
```
Comment (anirudh723) [inline src/components/RegistrationFormBody.tsx:11]:
```suggestion
  function generateName(name: string) {
     return field ? [field.name, name] : name;
  }
```

File: src/components/RegistrationFormBody.tsx [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+    if (field) {
+      return [field.name, name];
+    } else {
+      return name;
+    }
+  }
+
+  function generateFieldKey(key: string) {
```
Comment (anirudh723) [inline src/components/RegistrationFormBody.tsx:19]:
```suggestion
  function generateFieldKey(key: string) {
     return field ? [field.fieldKey, key] : key;
  }
```

File: src/containers/setContacts/ducks/types.ts [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+  school: string;
+  schoolYear: string;
+  allergies?: string;
+  diagnosis?: string;
+  medications?: string;
+  notes?: string;
+  photoRelease?: boolean;
+  profilePicture?: P;
+  dateOfBirth: D;
+}
```
Comment (anirudh723) [inline src/containers/setContacts/ducks/types.ts:50]: Lots of redundant fields. Consider using `Omit` https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys for creating GenericChild from GenericContact, omit the properties you don't need, and then simply add the ones that are new.

File: src/containers/setContacts/ducks/types.ts [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+
+export interface ContactInfo {
+  mainContact: Contact;
+  additionalContacts: AdditionalContact[];
+  children: Child[];
+  location: Location;
+}
+
+export type Contact = GenericContact<Date, string>;
+
```
Comment (anirudh723) [inline src/containers/setContacts/ducks/types.ts:60]: I think it's readable enough if you remove the extra white space

File: src/containers/setContacts/index.tsx (hunk 7 above)
Comment (anirudh723) [reply src/containers/setContacts/index.tsx:44]: I agree

File: src/containers/setContacts/index.tsx [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+  const registeringAsParticipatingFamily =
+    new URLSearchParams(location.search).get(
+      participatingFamilySearchQueryFlag,
+    ) != null;
+
+  if (asyncRequestIsComplete(setContacts) && registeringAsParticipatingFamily) {
+    protectedApiClient.makePFRequest();
+    history.push({
+      pathname: Routes.SIGNUP_CONFIRMATION,
+      search: participatingFamilySearchQuery,
```
Comment (anirudh723) [inline src/containers/setContacts/index.tsx:53]: 👍

File: src/utils/fileEncoding.ts [hunk 21]
```diff
@@ -19,14 +19,14 @@ export function convertToBase64String(file: File): Promise<string> {
   });
 }
 
-interface FileField {
+export interface FileField {
   file: File;
```
Comment (anirudh723) [inline src/utils/fileEncoding.ts:23]: Is there a need for this interface, if there's only one field?

File: src/App.tsx [hunk 22]
```diff
@@ -105,6 +109,16 @@ const App: React.FC = () => {
...
+                      <Route
+                        path={Routes.SET_CONTACTS}
+                        exact
+                        component={SetContacts}
+                      />
+                      <Route
+                        path={Routes.SIGNUP_CONFIRMATION}
+                        exact
+                        component={SignupConfirmation}
+                      />
```
Comment (anirudh723) [inline src/App.tsx:121]: I feel like we should just create a `routes` object, mapping a Route to a component. And simply just map through its entries. Then we can just add entries to the object. @jackblanc thoughts?

File: src/api/protectedApiClient.ts (hunk 1 above)
Comment (anirudh723) [reply src/api/protectedApiClient.ts:18]: Seems like this really wouldn't need to be in redux. Not too sure though. @jackblanc thoughts?

File: src/api/protectedApiClient.ts [hunk 23]
```diff
@@ -13,13 +14,16 @@ export interface ProtectedApiClient {
...
+  readonly getContactInfo: () => Promise<ContactInfo>;
+  readonly setContactInfo: (request: ContactInfo) => Promise<void>;
 }
 
 export enum ProtectedApiClientRoutes {
   CHANGE_PASSWORD = '/api/v1/protected/user/change_password',
   REQUEST_STATUSES = '/api/v1/protected/requests/status',
   MAKE_PF_REQUEST = 'api/v1/protected/requests',
   USER = '/api/v1/protected/user',
+  CONTACT_INFO = '/api/v1/protected/user/contact_info',
```
Comment (anirudh723) [inline src/api/protectedApiClient.ts:26]: ⛏ 
```
const BASE = '/api/v1/protected';
export enum ProtectedApiClientRoutes {
   CHANGE_PASSWORD = `${BASE}/user/change_password`,
   REQUEST_STATUSES = `${BASE}/requests/status`,
   ....
}
```
Removes a teeny tiny bit of risk of mistyping the route. Totally optional tho

File: src/api/protectedApiClient.ts (hunk 2 above)
Comment (anirudh723) [reply src/api/protectedApiClient.ts:44]: a "Completed" action with an error response does not sound like a good idea. I would remove the error catching if it is not possible to catch here

File: src/api/protectedApiClient.ts [hunk 24]
```diff
@@ -40,6 +44,16 @@ const deactivateAccount = (): Promise<void> => {
...
 };
 
+const getContactInfo = (): Promise<ContactInfo> => {
+  return AppAxiosInstance.get(ProtectedApiClientRoutes.CONTACT_INFO).then(
+    (res) => res.data,
+  );
+};
+
+const setContactInfo = (request: ContactInfo): Promise<void> => {
+  return AppAxiosInstance.put(ProtectedApiClientRoutes.CONTACT_INFO, request);
```
Comment (anirudh723) [inline src/api/protectedApiClient.ts:54]: I actually like this. No need for a then/catch when its returning a `Promise<void>`. Can you quickly fix that for the others ones too(`makePFRequest`, `changePassword `, `deactivateAccount`, etc)?

File: src/components/ChildFormFragment.tsx (hunk 3 above)
Comment (anirudh723) [reply src/components/ChildFormFragment.tsx:16]: Also, is there a way to get the asterisk on the right side of the label? I think that is more standard. Not sure if it is simple enough with AntD forms...

File: src/components/ChildFormFragment.tsx [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+      >
+        <TextArea rows={3} placeholder="Other Notes" />
+      </Form.Item>
+
+      <Form.Item
+        name={[field.name, 'profilePicture']}
+        fieldKey={[field.fieldKey, 'profilePicture']}
+        label="Upload Profile Picture"
+      >
+        <Dragger multiple={false} beforeUpload={() => false}>
```
Comment (anirudh723) [inline src/components/ChildFormFragment.tsx:122]:
```suggestion
        <Dragger 
          multiple={false} 
          beforeUpload={() => false}
        >
```

File: src/containers/setContacts/ducks/types.ts [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+export type Child = GenericChild<Date, string>;
+
+export interface AdditionalContact extends Contact {
+  shouldSendEmails?: boolean;
+}
+export interface Location {
```
Comment (anirudh723) [inline src/containers/setContacts/ducks/types.ts:66]: ⛏ Could enter a space before this type

File: src/containers/setContacts/index.tsx [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+  const dispatch = useDispatch();
+  const history = useHistory();
+  const location = useLocation();
+  useEffect(() => {
+    dispatch(getContactInfo());
+  }, [dispatch]);
+  const registeringAsParticipatingFamily =
+    new URLSearchParams(location.search).get(
+      participatingFamilySearchQueryFlag,
+    ) != null;
```
Comment (anirudh723) [inline src/containers/setContacts/index.tsx:47]: I think `URLSearchParams` includes a `has` function that would make this a bit simpler. I would write a test for this if you haven't already done so.

File: src/containers/setContacts/index.tsx (hunk 11 above)
Comment (anirudh723) [reply src/containers/setContacts/index.tsx:147]: We gotta get rid of Momentjs man...

**Review by anirudh723 — COMMENTED** [review]
Looks pretty good so far. There are a few things that could be fixed up, but the design looks good.

File: src/components/ChildFormFragment.tsx (hunk 25 above)
Comment (rymaju) [reply src/components/ChildFormFragment.tsx:122]: I think prettier disagrees so I wont make this change just based on that, I do think it might look better that way though!

File: src/components/LocationFormFragment.tsx (hunk 15 above)
Comment (rymaju) [reply src/components/LocationFormFragment.tsx:20]: Because the type of this input isnt numeric it should always be a string

File: src/components/LocationFormFragment.tsx (hunk 14 above)
Comment (rymaju) [reply src/components/LocationFormFragment.tsx:46]: Yeah, I think we need to relax the validation

File: src/containers/setContacts/ducks/types.ts (hunk 18 above)
Comment (rymaju) [reply src/containers/setContacts/ducks/types.ts:50]: Good idea, I think we can avoid that for simplicity and make a AbstractContactInfo or something that both contacts and child extend

File: src/utils/fileEncoding.ts (hunk 21 above)
Comment (rymaju) [reply src/utils/fileEncoding.ts:23]: It does just because it describes the structure of the AntD file field and its used in multiple places

File: src/api/protectedApiClient.ts (hunk 2 above)
Comment (jackblanc) [reply src/api/protectedApiClient.ts:44]: I'm pretty sure after catching, we still return a rejected promise to the thunk

File: src/App.tsx (hunk 22 above)
Comment (jackblanc) [reply src/App.tsx:121]: Oh that's interesting, I like that. We could also try to store a concept of like "minimum privilege level" required to access the page.

File: src/api/protectedApiClient.ts (hunk 1 above)
Comment (jackblanc) [reply src/api/protectedApiClient.ts:18]: Doesn't necessarily have to use Redux. The alternative is local `useState` using an `AsyncRequest` object. I was not very clear at the beginning of this semester that Redux doesn't always have to be used.

File: src/api/protectedApiClient.ts (hunk 23 above)
Comment (jackblanc) [reply src/api/protectedApiClient.ts:26]: might as well.

File: src/components/ChildFormFragment.tsx (hunk 3 above)
Comment (jackblanc) [reply src/components/ChildFormFragment.tsx:16]: Who is this a question for haha

File: src/components/LocationFormFragment.tsx (hunk 5 above)
Comment (jackblanc) [reply src/components/LocationFormFragment.tsx:45]: yes.

Comment (jackblanc) [reply src/components/LocationFormFragment.tsx:45]: Easiest way is to define these in `utils` as constants and then just reference the RegExp object here and in isolated unit tests

File: src/containers/setContacts/ducks/actions.ts [hunk 28]
```diff
@@ -0,0 +1,15 @@
+import { genericAsyncActions } from '../../../utils/asyncRequest';
+import { ContactInfo } from './types';
+
+export const contacts = genericAsyncActions<ContactInfo, any>();
+export const setContacts = genericAsyncActions<void, any>();
```
Comment (jackblanc) [inline src/containers/setContacts/ducks/actions.ts:5]: Does the setContacts API return the ContactInfo in the response? Either way, I think we can get away with one action, probably.

File: src/containers/setContacts/ducks/thunks.ts [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+      });
+  };
+};
+
+export const resetSetContactState = (): ContactsThunkAction<void> => {
+  return (dispatch, getState, _) => {
+    dispatch(setContacts.loading());
```
Comment (jackblanc) [inline src/containers/setContacts/ducks/thunks.ts:39]: Shouldn't this be `notStarted()`?

File: src/containers/setContacts/ducks/thunks.ts [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+  request: ContactInfo,
+): ContactsThunkAction<void> => {
+  return (dispatch, getState, { protectedApiClient }) => {
+    dispatch(contacts.loading());
+    dispatch(setContacts.loading());
+    return protectedApiClient
+      .setContactInfo(request)
+      .then(() => {
+        dispatch(contacts.loaded(request));
+        dispatch(setContacts.loaded());
```
Comment (jackblanc) [inline src/containers/setContacts/ducks/thunks.ts:28]: I think the best solution here is to have the API return ContactInfo, and use one action `contacts` to track both interactions.

File: src/containers/setContacts/ducks/types.ts (hunk 18 above)
Comment (jackblanc) [reply src/containers/setContacts/ducks/types.ts:50]: I agree, abstract is good

File: src/containers/setContacts/ducks/types.ts [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+import { ThunkAction } from 'redux-thunk';
+import { ProtectedApiExtraArgs } from '../../../api/protectedApiClient';
+import { C4CState } from '../../../store';
+import { AsyncRequest } from '../../../utils/asyncRequest';
+import { FileField } from '../../../utils/fileEncoding';
+import { ContactsActions } from './actions';
+
+export interface ContactsReducerState {
+  readonly contacts: AsyncRequest<ContactInfo, any>;
+  readonly setContacts: AsyncRequest<void, any>;
```
Comment (jackblanc) [inline src/containers/setContacts/ducks/types.ts:11]: Probably can get away with one.

File: src/containers/setContacts/ducks/types.ts [hunk 32]
```diff
@@ -0,0 +1,85 @@
...
+  allergies?: string;
+  diagnosis?: string;
+  medications?: string;
+  notes?: string;
+  photoRelease?: boolean;
+  referrer?: string;
+  profilePicture?: P;
+}
+
+interface GenericChild<D, P> {
```
Comment (jackblanc) [inline src/containers/setContacts/ducks/types.ts:37]: Why are generics required?

File: src/containers/setContacts/index.tsx (hunk 7 above)
Comment (jackblanc) [reply src/containers/setContacts/index.tsx:44]: yes

File: src/containers/setContacts/index.tsx (hunk 8 above)
Comment (jackblanc) [reply src/containers/setContacts/index.tsx:57]: maybe a jsdoc :)

File: src/containers/setContacts/index.tsx (hunk 9 above)
Comment (jackblanc) [reply src/containers/setContacts/index.tsx:73]: Wait, won't this always throw an error? the AsyncRequest `contacts` will not be complete when onFinish is called?

File: src/containers/setContacts/index.tsx (hunk 11 above)
Comment (jackblanc) [reply src/containers/setContacts/index.tsx:147]: agreed

**Review by jackblanc — COMMENTED** [review]
This looks great, definitely some cleanup could be done in the main setContacts container. I think we can get away with one AsyncRequest field as well.

File: src/App.tsx (hunk 22 above)
Comment (onb6) [reply src/App.tsx:121]: That's a good idea! Definitely outside the scope of this PR, but could make routing cleaner going forward.

File: src/components/ChildFormFragment.tsx [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+          <p>Drag and Drop Image File to Upload (.jpeg, .png)</p>
+          <u>Or Browse Your Computer</u>
+        </Dragger>
+      </Form.Item>
+    </>
+  );
+};
+
+export default ChildFormFragment;
```
Comment (onb6) [inline src/components/ChildFormFragment.tsx:131]: 👍  looks good! (other than the classNames which i agree don't makes sense)

File: src/components/ContactsForm.tsx [hunk 34]
```diff
@@ -0,0 +1,125 @@
...
+                    >
+                      Remove
+                    </Button>
+                    <Divider />
+                  </div>
+                ))}
+
+                <Button
+                  onClick={() => {
+                    add();
```
Comment (onb6) [inline src/components/ContactsForm.tsx:73]: 👍

File: src/containers/setContacts/ducks/thunks.ts [hunk 35]
```diff
@@ -0,0 +1,41 @@
...
+    return protectedApiClient
+      .getContactInfo()
+      .then((response: ContactInfo) => {
+        dispatch(contacts.loaded(response));
+      })
+      .catch((error: any) => {
+        dispatch(contacts.failed(error));
+      });
+  };
+};
```
Comment (onb6) [inline src/containers/setContacts/ducks/thunks.ts:16]: 👍

**Review by onb6 — APPROVED** [review]
Apart from the changes already suggested, looks good!

File: src/containers/setContacts/index.tsx (hunk 9 above)
Comment (rymaju) [reply src/containers/setContacts/index.tsx:73]: contacts should be finished but setContacts will be pending or resolved

File: src/components/ContactsForm.tsx [hunk 36]
```diff
@@ -0,0 +1,125 @@
...
+  onFinish,
+  initialValues,
+  isParticipatingFamily,
+}) => {
+  return (
+    <FormContainer>
+      <Helmet>
+        <title>Set Profile Information</title>
+      </Helmet>
+      {isParticipatingFamily ? (
```
Comment (vrautela) [inline src/components/ContactsForm.tsx:29]:
```suggestion
      {isInPFSignupFlow ? (
```

File: src/components/ContactsForm.tsx [hunk 37]
```diff
@@ -0,0 +1,125 @@
...
+        <ConfirmationMessage
+          title="UPDATE PROFILE INFORMATION"
+          message="Contacts, Children, and Profile Information"
+          details="Tell us about your children and/or any other members you are registering under this account."
+        />
+      )}
+
+      <Form onFinish={onFinish} layout="vertical" initialValues={initialValues}>
+        <Title level={5}>Main Contact Information</Title>
+        <RegistrationFormBody isMainContact />
```
Comment (vrautela) [inline src/components/ContactsForm.tsx:45]: ⛏️ I don't like having the default value prop stuff

File: src/components/ContactsForm.tsx [hunk 38]
```diff
@@ -0,0 +1,125 @@
...
+        </Form.List>
+        <Button type="primary" htmlType="submit">
+          Submit
+        </Button>
+      </Form>
+    </FormContainer>
+  );
+};
+
+export default ContactsForm;
```
Comment (vrautela) [inline src/components/ContactsForm.tsx:125]: 👍

File: src/components/RegistrationFormBody.tsx [hunk 39]
```diff
@@ -0,0 +1,156 @@
...
+        fieldKey={generateFieldKey('phoneNumber')}
+        label="Phone Number"
+        rules={[
+          {
+            required: true,
+            pattern: new RegExp(/^[^0-9]*(?:(\d)[^0-9]*){10}$/),
+            message: 'Please input a valid phone number',
+          },
+        ]}
+        style={{ display: 'inline-block', width: 'calc(50% - 8px)' }}
```
Comment (vrautela) [inline src/components/RegistrationFormBody.tsx:91]: 👁️ 👄 👁️ inline styling

**Review by vrautela — APPROVED** [review]
